cql3: deprecate and start replacing cql3::util::maybe_quote

nyh opened the pull request:
cql3::util::maybe_quote() is a utility function formatting an identifier
name (table name, column name, etc.) that needs to be embedded in a CQL
statement - and might require quoting if it contains non-alphanumeric
characters, uppercase characters, or a CQL keyword.

maybe_quote() made a valiant effort to only quote the identifier name
if neccessary - e.g., a lowercase name does not need quoting. Or, so
we thought. I now consider that effort to be misdirected: It turns out
that even lowercase names sometimes need to be quoted - when they are
CQL keywords they cannot be used without quoting. This can cause
problems for code that wants to generate CQL statements, such as the
materialized-view problem in issue #9450.

So this patch concedes that maybe_quote() is problematic - it doesn't
always quote when necessary - and deprecates it. Instead it adds a new
function - quote() - which unconditionally quotes an identifier name.
It is always legal to add quotes even when not necessary.

I did not want to replace all uses of maybe_quote() with quote() because
I am not sure how to test all the places it is used. So I changed only
the one place in the code which is needed to fix issue #9450 - the
column identifier's to_cql_string() function. Probably eventually all
other calls to maybe_quote() should be reviewed, tests added for the
CQL-keyword case, and the call to maybe_quote() should be replaced by
quote().

Fixes #9450

Signed-off-by: Nadav Har'El <nyh@scylladb.com>

nyh commented:
I sent this patch to the mailing list in October 2021 - see thread https://groups.google.com/g/scylladb-dev/c/G19lxJASTXs/m/NFoxA117CAAJ - where it was reviewed by @avikivity and then lost. So moving it to a pull-request where it won't get lost. 

Although this patch fixes a real bug, it also "pushes" the idea that maybe_quote shouldn't ever be used, and there is no harm of always quoting names in CQL. What Avi didn't like about _that_ idea is that it can make some user-visible printouts - e.g., perhaps, schemas - uglier with all the unnecessary quotes. Avi suggested leaving maybe_quote() but having it run the parser to check whether an unquoted name matches IDENT. I was worried this might be slow but Avi said it was acceptable. 

So perhaps we should go in that way - leave `maybe_quote()` as is (don't deprecate it, don't introduce a new `quote()`), just add to it the parser verification as another reason to do quoting.

One not-very-nice feature about the alternative implementation is that long-living CQL statements (like schemas saved to a file)  will break if the parser changes in the future and new keywords are added - so that a name that previously worked unquoted will no longer work unquoted. This is, however, not relevant to short-lived statements like the one in issue 9450.

avikivity commented:
I think we should make an effort to make it work. Perhaps run the identifier via the parser for the `cident` production, and if it fails, quote it.

OTOH, this fails if we add a reserved identifier later, and then load an old schema with a new database. This should be very rare, but it can happen.

nyh commented:
Converted this PR to draft, and I'll replace it by one that keeps `maybe_quote()` but adds the feature of recognizing reserved identifiers and quoting them too.

avikivity commented:
Like I mentioned, it has a downside (lack of forward compatibility). It's not clear that it's a better solution. It depends on what we use maybe_quote for.

nyh commented:
> Like I mentioned, it has a downside (lack of forward compatibility). It's not clear that it's a better solution. It depends on what we use maybe_quote for.

We use it for a lot of different things... The use that sparked this case was save the restriction as a string in materialized view's definition. This can indeed suffer from that "forward compatibility" problem of newly introduced keywords breaking views with that name.

I can also have the cake and eat it too: I can improve maybe_quote() (to also look at the list of keywords and quote them), and at the same time, also use unconditional quote() when the result is not user visible and doesn't need to be pretty (like in the issue this patch is trying to solve - the saved restriction of a materialized view). The problem is that we might have convenience functions that call maybe_quote() - like the `column_identifier::to_cql_string()` I fixed in this patch - that may perhaps be called from different places for different purposes. So would I split to_cql_string() into two, one calling maybe_quote() and one quote()? It sounds like a mess...

nyh commented:
Instead of this PR, I sent two patches to the mailing list:

1. `cql3: fix cql3::util::maybe_quote() for keywords` - fixing maybe_quote() to quote reserved keywords (but not other lowercase identifiers), and fixing issue #9450.
2. `cql3: add a cql3::util::quote() function` - adding a new unconditional `quote()` function. This function is not yet used anywhere - it's not necessary for solving #9450. But we can consider in the future to use it in some, most or all places that today use maybe_quote().

nyh commented:
I'm closing this old PR. Instead of removing maybe_quote(), we now have both maybe_quote() and quote(), where the latter always quotes, and code has its choice of which one to use.
